src/engine/BMBtnSkillEcho.php

```diff
@@ -9,41 +9,56 @@
  * This class currently supports the special skills of Echo
  */
 class BMBtnSkillEcho extends BMBtnSkill {
-    public static $hooked_methods = array('load_buttons');
+    public static $hooked_methods = array('specify_recipes');
 
-    public static function load_buttons(array $args) {
+    public static function specify_recipes(array $args) {
         $areAllArgsPresent =
-            array_key_exists('name', $args) &&
-            array_key_exists('recipe', $args) &&
-            array_key_exists('oppname', $args) &&
-            array_key_exists('opprecipe', $args);
+            array_key_exists('button', $args) &&
+            array_key_exists('oppbutton', $args);
 
         if (!$areAllArgsPresent) {
-            throw new LogicException('load_buttons die hook is missing required input arguments');
+            throw new LogicException('specify_recipes die hook is missing required input arguments');
         }
 
-        if (empty($args['name'])) {
+        $button = $args['button'];
+        $oppbutton = $args['oppbutton'];
+
+        if (is_null($oppbutton)) {
+            return;
+        }
+
+        if (!($button instanceof BMButton) ||
+            !($oppbutton instanceof BMButton)) {
+            throw new LogicException('specify_recipes requires two BMButton input arguments');
+        }
+
+        if (empty($button->name)) {
             throw new LogicException('Button name may not be empty.');
         }
 
         // only copy recipe if the opponent button exists
-        if (empty($args['oppname'])) {
+        if (empty($oppbutton->name)) {
             return;
         }
 
-        $newRecipe = $args['recipe'];
+        $newRecipe = $button->recipe;
 
         // copy opponent's recipe only if Echo hasn't yet got a recipe
-        if ('' == $args['recipe']) {
-            if ('' == $args['opprecipe']) {
-                // opponent's button recipe is empty, which only occurs with Echo and Zero
-                // thus choose a default recipe
+        if ('' == $button->recipe) {
+            $newRecipe = $oppbutton->recipe;
+
+            if ('' == $newRecipe &&
+                (
+                    ('Echo' == $oppbutton->name) ||
+                    ('Zero' == $oppbutton->name)
+                )
+               ) {
```

Comment:
Why do Echo and Zero need to be hardcoded now, when they didn't used to be?  Can't you look for the presence of BMBtnSkillEcho in the opponent button somehow?